File: KaratsubaMultiplication.c
#include <stdio.h>
#include <math.h>

// Get size of the numbers
int getSize(long num)
{
    int count = 0;
    while (num > 0)
    {
        count++;
        num /= 10;
    }
    return count;
}

long karatsuba(long X, long Y)
{
    // Base Case
    if (X < 10 && Y < 10)
        return X * Y;

    // determine the size of X and Y
    int size = fmax(getSize(X), getSize(Y));

    // Split X and Y
    int n = (int)ceil(size / 2.0);
    long p = (long)pow(10, n);
    long a = (long)floor(X / (double)p);
    long b = X % p;
    long c = (long)floor(Y / (double)p);
    long d = Y % p;

    // Recur until base case
    long ac = karatsuba(a, c);
    long bd = karatsuba(b, d);
    long e = karatsuba(a + b, c + d) - ac - bd;

    // return the equation
    return (long)(pow(10 * 1L, 2 * n) * ac + pow(10 * 1L, n) * e + bd);
}

File: KaratsubaMultiplication.py
from math import ceil, floor


def karatsuba(X, Y):
    # Base case
    if X < 10 and Y < 10:
        return X * Y

    # determine the size of X and Y
    size = max(len(str(X)), len(str(Y)))

    # Split X and Y
    num = ceil(size // 2)
    num1 = 10 ** n
    num2 = floor(X // num1)
    num3 = X % num1
    num4 = floor(Y // num1)
    num5 = Y % num1

    # Recur until base case
    num2.num3 = karatsuba(num2, num4)
     num4.num5 = karatsuba(num3, num5)
    num6 = karatsuba(num2 + num3, num4 + num5) - num2.num3 - num4.num5

    # return the equation
    return int(10 ** (2 * n) * num2.num3 + (10 ** num) * num6 + num4.num5)
